### PR TITLE
Support running HttpApi locally

### PR DESCRIPTION
**Pull request by pyk**
*Issue #, if available:* #1641

*Why is this change necessary?*
This changes is necessary in order to make `HttpApi` event type work with `sam local start-api`.

*How does it address the issue?*

1. Currently `sam` doesn't support `HttpApi` because it only extract resource with type `AWS::Serverless::Function` and `AWS::Serverless::Api`, this changes add `AWS::Serverless::HttpApi` to be extracted from the `template.yaml`
2. Construct [event v2.json](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-integrations-lambda.html) from flask request

*What side effects does this change have?*

*Did you change a dependency in `requirements/base.txt`?*
    *If so, did you run `make update-reproducible-reqs`*
No

*Checklist:*

- [ ] Write Design Document ([Do I need to write a design document?](https://github.com/awslabs/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.rst#design-document))
- [x] Write unit tests
- [x] Write/update functional tests
- [x] Write/update integration tests
- [x] `make pr` passes
- [x] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by pyk** [comment]
Example template:

```
Resources:
    HTTPAPIGatewayLogGroup:
        Type: AWS::Logs::LogGroup
        Properties:
            LogGroupName: HTTPAPIGatewayLogGroup
            RetentionInDays: 1

    HTTPAPIGateway:
        Type: AWS::Serverless::HttpApi
        Properties:
            AccessLogSettings:
                DestinationArn: !GetAtt HTTPAPIGatewayLogGroup.Arn
                Format: '{"requestId": "$context.requestId", "userAgent": "$context.identity.userAgent", "httpMethod": "$context.httpMethod", "path": "$context.path", "status": "$context.status", "protocol": "$context.protocol", "integrationStatus": "$context.integrationStatus", "responseLatency": "$context.responseLatency" }'
            FailOnWarnings: true

    LambdaLoginPage:
        Type: AWS::Serverless::Function
        Properties:
            FunctionName: LoginPage
            Description: >
                A function to handle /login
            MemorySize: 128
            Timeout: 3
            CodeUri: app/
            Handler: controllers.login_page
            Runtime: python3.8
            Events:
                HTTPMethodGET:
                    Type: HttpApi
                    Properties:
                        ApiId: !Ref HTTPAPIGateway
                        Path: /login
                        Method: get
                        TimeoutInMillis: 3000
```

To run `HttpApi` locally:

```
sam local start-api
```

For example on windows:

```
(samcli-dev) PS C:\Users\bayu\pyk\streamers-serverless> .\samcli-dev\Scripts\sam.exe local start-api
Mounting LambdaLoginPage at http://127.0.0.1:3000/login [GET]
You can now browse to the above endpoints to invoke your functions. You do not need to restart/reload SAM CLI while working on your functions, changes will be reflected instantly/automatically. You only need to restart SAM CLI if you update your AWS SAM template
2020-04-25 21:47:40  * Running on http://127.0.0.1:3000/ (Press CTRL+C to quit)
2020-04-25 21:47:43 127.0.0.1 - - [25/Apr/2020 21:47:43] "GET / HTTP/1.1" 403 -
2020-04-25 21:47:43 127.0.0.1 - - [25/Apr/2020 21:47:43] "GET /favicon.ico HTTP/1.1" 403 -
Invoking controllers.login_page (python3.8)

Fetching lambci/lambda:python3.8 Docker container image......
Mounting C:\Users\bayu\pyk\streamers-serverless\.aws-sam\build\LambdaLoginPage as /var/task:ro,delegated inside runtime container
START RequestId: af2687b6-7ca5-1794-302d-1faa63af6e0c Version: $LATEST
END RequestId: af2687b6-7ca5-1794-302d-1faa63af6e0c
REPORT RequestId: af2687b6-7ca5-1794-302d-1faa63af6e0c     Init Duration: 332.88 ms        Duration: 4.33 ms       Billed Duration: 100 ms Memory Size: 128 MB     Max Memory Used: 24 MB  
2020-04-25 21:47:55 127.0.0.1 - - [25/Apr/2020 21:47:55] "GET /login HTTP/1.1" 200 -
2020-04-25 21:47:55 127.0.0.1 - - [25/Apr/2020 21:47:55] "GET /favicon.ico HTTP/1.1" 403 -
```

File: samcli/local/apigw/local_apigw_service.py
Comment (jnodorp) [inline samcli/local/apigw/local_apigw_service.py:474]: I think this unused variable is the only reason for the failing check.

```suggestion
```

Comment (pyk) [reply samcli/local/apigw/local_apigw_service.py]: @jnodorp thanks!

**Comment by pyk** [comment]
I'm waiting a review from the maintainers.

I saw many peoples are wanted this feature #1641 , the temporary solution is to install this branch using the following command:

```
pip install git+https://github.com/pyk/aws-sam-cli.git@develop
```

Then run:

```
sam local start-api
```

Make sure to uninstall the previous sam-cli.

**Comment by rowanu** [comment]
Can confirm that this is working locally for me 👍 thanks again @pyk

**Comment by jo5ef** [comment]
> Can confirm that this is working locally for me  thanks again @pyk

me too

**Comment by WinterYukky** [comment]
Hello @pyk
I thank you very much.

I'm really looking forward to using HttpApi. I tried it as @pyk provided a temporary solution. However, CORS was not available. I have created a pull request so that CORS can be configured with HttpApi, so please check.

**Comment by picosam** [comment]
I’m curious to know what’s blocking this merge? It keeps becoming ready then again goes into conflict mode.

**Comment by pyk** [comment]
@picosam It seems timeout(?) causing `make pr` test failed in 2 environments:

Annotation 2020-06-21 142913

I hope I can rerun the test to validate this.

**Comment by mistaecko** [comment]
Can you push an empty commit? `git commit --allow-empty -m "trigger build"`

Looking at the build logs the two builds failed at different steps without any apparent error message. Could this simply be a build server instability? That said, looking at the commits on this PR it seems that 10 out of 12 builds have failed so far.
Someone with experience on developing for sam-cli might be able to tell if that does indeed indicate an issue with the code or not.

**Comment by pyk** [comment]
@mistaecko done 👍

**Comment by picosam** [comment]
Thank you @pyk, I sure hope someone's going to review this and merge it once and for all.

**Comment by svenemtell** [comment]
@pyk Tried the temporary solution
> pip install git+https://github.com/pyk/aws-sam-cli.git@run-httpapi-locally

and noticed that Payload format version 1.0* always seems to be used when running using `sam local start-api`.
Is this expected?

*) https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-integrations-lambda.html

**Comment by sebsto** [comment]
Can someone review and approve this ?

**Comment by sebsto** [comment]
I confirm local installation described at https://github.com/awslabs/aws-sam-cli/pull/1942#issuecomment-644063266 works for me

**Comment by sebsto** [comment]
> @pyk Tried the temporary solution
> 
> > pip install git+https://github.com/pyk/aws-sam-cli.git@run-httpapi-locally
> 
> and noticed that Payload format version 1.0* always seems to be used when running using `sam local start-api`.
> Is this expected?
> 
> *) https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-integrations-lambda.html

I confirm the payload version 1.0 seems to be used; even with HttpApi.
Looks like the PR is missing a new payload definition in https://github.com/awslabs/aws-sam-cli/tree/develop/samcli/lib/generated_sample_events/events/apigateway

**Comment by picosam** [comment]
Still desperately waiting for this to be merged.

**Comment by logemann** [comment]
So this really useful PR sits idle since April without any feedback from AWS ? This is really kind of demotivating for developers to even start contributing.

**Comment by picosam** [comment]
@awsjeffg is there anything we can maybe do to help out?

**Comment by awsjeffg** [comment]
@picosam nothing currently but thanks for the nudge. We're evaluating this and other popular github issues for sam and sam cli as part of a post-1.0 review process. What I would say about this type of issue generally is that we ( AWS ) need to be crisp in terms of what we support in emulation locally and why, as there are potential issues with providing something for local development that does not behave like the hosted service in subtle ways. This issue and others like it are in this grey area.

**Comment by Ozzee** [comment]
As a developer I would be extremely happy with a version that works even on just a small subset of the real features. Most of the use cases I see are simple api endpoints that do not use any advanced features. The local emulation does not need to be a perfect replica of the actual hosted service.

Our current solution is keeping both http api and api gateway code in the template and switching between them by commenting and uncommenting parts of the template. **Any** minimal support for http api would be better than this...

**Comment by LaurenceGA** [comment]
On the flipside, according to [this earlier comment](https://github.com/aws/aws-sam-cli/pull/1942#issuecomment-654092365), if I understand correctly, this PR only uses payload version 1.0 no matter what though.
The latest payload version can barely be called an advanced feature that's missing I think...
So surely this PR is not even mergeable right now...?
I appreciate AWS, as they are, being thorough in making sure that the local emulation is being brought up to a good standard.

Hopefully this is high priority to get this one up to scratch though, because it would ease a lot of people's pain trying to develop for what AWS say are the latest tools - HTTP APIs aren't in beta anymore, they should have proper tooling support so that we can confidently adopt them...

**Comment by WinterYukky** [comment]
Hello @pyk
I sent a PR to address the issue of "payload Format Version", so please check.

**Comment by Hobbit71** [comment]
I can see merging is blocked because of conflicts. So desperate for this change. I would attempt to resolve the merge conflicts if I knew anything about Python. 

AWS - this is key functionality, can't you get involved and progress this.

**Comment by awsjeffg** [comment]
@Hobbit71 thanks for the feedback - we're looking into taking this change and adding some additional changes to provide this feature, hopefully in the next couple of months.

**Comment by moelasmar** [comment]
@pyk 
There are couple of PRs (https://github.com/pyk/aws-sam-cli/pull/5, https://github.com/pyk/aws-sam-cli/pull/6) on your repo that are related to this change. Could you please merge them, and update this PR so we can merge it to our repo.

**Comment by pyk** [comment]
@moelasmar Done 👌

**Comment by pyk** [comment]
@moelasmar I make a few changes, please review it 🙇‍♂️

**Comment by wchengru** [comment]
@pyk Can you please also merge [pyk#7]([url](https://github.com/pyk/aws-sam-cli/pull/7)) to your branch before we merging it? Thanks!

**Review by wchengru — APPROVED** [review]

**Review by moelasmar — APPROVED** [review]

File: samcli/local/events/api_event.py
Comment (awood45) [inline samcli/local/events/api_event.py:284]: Possibly legacy carry-over, but why always the same request ID here? Can we just generate a new valid request ID each time, in case someone is locally testing where they check for duplicated request IDs?

This shouldn't block launch if we're copying logic from REST API.

**Review by awood45 — CHANGES_REQUESTED** [review]
One additional point is that I'd like to see a test where we have a REST API and HTTP API in the same template, and that we handle that correctly. We should, and the use case is uncommon, but we should ensure it works properly and that we do not regress in behavior for that case.

**Review by awood45 — APPROVED** [review]
My earlier comments stand, and will be addressed in a fast-follow PR before the next release.

**Comment by awood45** [comment]
Thanks so much @pyk for the contribution!

**Comment by awsjeffg** [comment]
Yes @pyk thanks so much for contributing, and for your patience!

**Comment by pyk** [comment]
you are welcome guys! 🙇

**Comment by tsndr** [comment]
So will it be part of the next release? :)

**Comment by awsjeffg** [comment]
Yes!

**Comment by svenemtell** [comment]
It would be great to also have the accompanying #2334 merged!

**Comment by awsjeffg** [comment]
@svenemtell I believe we've provided v2 payload support by merging another contribution from @WinterYukky that was merged into @pyk 's PR:

https://github.com/aws/aws-sam-cli/pull/1942/commits/25f0169132e7e6085aedf3111deea824cfda9f15

( unless I'm misunderstanding something )

If you are testing this new feature once SAM cli v.1.7 is released and find issues with payload versions, please file a new issue and mention me. We're hoping for EOD today for this release ( it was supposed to be yesterday actually but we had an internal issue and had to delay for a day )

**Comment by awsjeffg** [comment]
@svenemtell  and, thanks for contributing that PR and sorry for taking this different path in getting these contributions merged, it ended up being the simplest thing to do.

**Comment by awsjeffg** [comment]
And we're live! I know many people are excited about this new feature - please test it out and if you run into issues, please file and issue and @ me.

**Comment by tsndr** [comment]
There's just one thing I don't really know if it's a bug or intentional. When I run my HttpApi locally and `console.log` the event object my `routeKey` is `/auth/login:POST`, but when I do the same thing on AWS Lambda it says `POST /auth/login`.

**Comment by awsjeffg** [comment]
@tsndr log that bug! Seems like it could be a shallow fix. Does anything break as a result?

**Comment by tsndr** [comment]
In my case it causes my "routing" to break. I'm using one lambda function to handle multiple endpoints, so I I'm doing this:

```
switch (event.routeKey) {
    case 'POST /auth/login':
        response = await require('./controllers/AuthController').login(event)
        break
    case 'GET /auth/user':
        response = await require('./controllers/AuthController').user(event)
        break
    case 'POST /auth/register':
        response = await require('./controllers/AuthController').register(event)
        break
}
```

Which might not be the cleanest solution, but it works for me.

**Comment by therealjeffg** [comment]
@tsndr it's still a behaviour difference and we should try to address it. Can you please file an issue if you have time? And @ me so I get an email reminder.

**Comment by awsjeffg** [comment]
> In my case it causes my "routing" to break. I'm using one lambda function to handle multiple endpoints, so I'm doing this:
> 
> ```
> switch (event.routeKey) {
>     case 'POST /auth/login':
>         response = await require('./controllers/AuthController').login(event)
>         break
>     case 'GET /auth/user':
>         response = await require('./controllers/AuthController').user(event)
>         break
>     case 'POST /auth/register':
>         response = await require('./controllers/AuthController').register(event)
>         break
> }
> ```
> 
> Which might not be the cleanest solution, but works for me.

This issue is being worked on in #2345

**Pull request merged**
